File: aditya sarkaar/ai_lab/lab 5/map_coloring.pl
% Australia Map Colouring Problem (Constraint Satisfaction)
% Regions: WA, NT, Q, NSW, V, SA, T
% Domain : {red, green, blue}  (encoded as 1, 2, 3)
% Constraint: adjacent regions must have different colours.

:- use_module(library(clpfd)).

% solve(-Colouring)
% Colouring is a list Region-Colour where Colour is in 1..3.
solution(Colouring) :-
    Colouring = [
        wa-WA, nt-NT, q-Q, nsw-NSW, v-V, sa-SA, t-T
    ],
    Colours = [WA, NT, Q, NSW, V, SA, T],
    Colours ins 1..3,

    % Adjacency constraints: neighbouring regions must differ.
    WA #\= NT, WA #\= SA,
    NT #\= Q,  NT #\= SA,
    Q  #\= NSW, Q  #\= SA,
    NSW #\= V, NSW #\= SA,
    V #\= SA,

    label(Colours).

% colour_name(+Code, -Name)
colour_name(1, red).
colour_name(2, green).
colour_name(3, blue).

% print_solution(+Colouring)
print_solution(Colouring) :-
    write('Australia Map Colouring Solution:'), nl,
    forall(
        member(Region-Code, Colouring),
        ( colour_name(Code, Name),
          format('  ~w = ~w~n', [Region, Name]) )
    ).

% all_solutions(-Solutions)
all_solutions(Solutions) :-
    findall(Colouring, solution(Colouring), Solutions).

File: aditya sarkaar/ai_lab/lab 5/map_coloring.py
# Australia Map Colouring Problem (Constraint Satisfaction)
# Regions: WA, NT, Q, NSW, V, SA, T
# Domain : {red, green, blue}
# Constraint: adjacent regions must have different colours.

from itertools import product

REGIONS = ["WA", "NT", "Q", "NSW", "V", "SA", "T"]
COLOURS = ["red", "green", "blue"]

# Adjacency list (undirected).
ADJACENT = {
    "WA": ["NT", "SA"],
    "NT": ["WA", "Q", "SA"],
    "Q": ["NT", "NSW", "SA"],
    "NSW": ["Q", "V", "SA"],
    "V": ["NSW", "SA"],
    "SA": ["WA", "NT", "Q", "NSW", "V"],
    "T": [],  # Tasmania has no mainland neighbours.
}


def is_valid(assignment):
    """Check whether every adjacency constraint is satisfied."""
    for region, colour in assignment.items():
        for neighbour in ADJACENT[region]:
            if assignment.get(neighbour) == colour:
                return False
    return True


def solve():
    """Find the first valid colouring by brute-force search."""
    for combo in product(COLOURS, repeat=len(REGIONS)):
        assignment = dict(zip(REGIONS, combo))
        if is_valid(assignment):
            return assignment
    return None


if __name__ == "__main__":
    solution = solve()
    if solution:
        print("Australia Map Colouring Solution:")
        for region in REGIONS:
            print(f"  {region} = {solution[region]}")
    else:
        print("No valid colouring found.")
